### ISSUE TITLE
The `nextPage`/`prevPage` seems to be the wrong way around?

### ISSUE THREAD

**Issue by helje5** [TootSDK/TootSDK#311]
So, I'm trying to download my bookmarks:
```swift
      var cursor: PagedInfo?
      repeat {
        let bookmarks = try await client
          .getTimeline(.bookmarks, pageInfo: cursor, limit: 10_000)
        print("Got bookmarks: #\(bookmarks.result.count)")
        cursor = bookmarks.nextPage
      }
      while cursor != nil
```
This only retrieves one batch.

This is because the `nextPage` property returns the _previous_ page link, and the other way around:
```swift
    /// Returns pagination object for the next page of results if one is available
    public var nextPage: PagedInfo? {
        if let sinceId = info.sinceId {
            return PagedInfo(sinceId: sinceId)
        }
        if let minId = info.minId { // <=== this one
            return PagedInfo(minId: minId)
        }
        return nil
    }
```

Those are the paging links of the API:
- https://mastodon.social/api/v1/bookmarks?limit=10000&max_id=3976996>; rel="next"
- https://mastodon.social/api/v1/bookmarks?limit=10000&min_id=4296133>; rel="prev"

Using `nextPage` picks the `prev` link. I wasn't 100% sure whether this is intentional, because technically the "next" items are "older" in the timeline. But the "next" for batching should probably still be the "next-batch".

P.S.: I think it would be better to use the actual next/prev links the API provided instead of parsing an recreating them.

**Comment by kkostov** [comment]
Yes, we consciously opted to invert prev and next but it was something we've struggled with a bit at the beginning. I'm happy to tweak it if we can find a way to improve it.

Here is some context:

`PageInfo` is generally used for two things:

* to enumerate large collections
* to seek items appearing before or after a given item (e.g. show me newer/older posts from post, or who has followed me after/before the given follower)

The solution we settled on is that `PagedResult.nextPage` will mean "more towards the top" of the list as they would be returned by the server. 

* to get a collection "starting from the top and going down", one would use `PagedResult.previousPage` to discover subsequent batches of results.
* To get results "more towards to top" from the current position, one would use `PagedResult.nextPage`.

Example: Given a timeline of posts with the latest posts on top. If you're holding the ID of a post, "nextPage" are newer posts and "previousPage" will provide older posts ("lower down" in the timeline). 

From TootSDK/app point of view, there is always a risk that "next"/"previous" can semantically mean different things depending on the instance giving the results. Not all server flavours have the same implementation for pagination and some also allow admins (and even users) to configure various views.

We have [some](https://github.com/TootSDK/TootSDK/blob/cd46f04a974c36013a30c29f45a8b38e082b8d82/Examples/swiftyadmin/Sources/swiftyadmin/Conversations/GetConversations.swift#L18) [examples](https://github.com/TootSDK/TootSDK/blob/cd46f04a974c36013a30c29f45a8b38e082b8d82/Examples/swiftyadmin/Sources/swiftyadmin/Account/GetFollowers.swift#L19) to try and show how to use the library in that sense.

**Comment by helje5** [comment]
Maybe it is a request for adding that info to the respective methods then :-)

**Comment by technicat** [comment]
I'm used to it now (or rather, I have a wrapper around it so I never see it), but now that I look at the Mastodon API again it is confusing for tootsdk to reverse the meanings. "next" matches nicely with "newer" for timelines, but maybe not getFollowers (or bookmarks for that matter). I would think of nextPage more like paging through a book. It's still confusing to think of new results showing up as previous, but blame Mastodon. This btw is my wrapper for retrieving a bunch of pages at once.

```swift
 func getPages<T: Pageable>(maxPages: Int = 5, query: PageQuery<T>)
    async throws
  {
    var count = 0
    var page = try await query(nil)
    while let prev = page.previousPage, count < maxPages {
      count += 1
      page = try await query(prev)
    }
  }

typealias PageQuery<T: Pageable> = (PagedInfo?) async throws -> PagedResult<[T]
```

**Comment by helje5** [comment]
Maybe it should just be different types if they have different meanings. A batch cursor type and a time based type. 🤷‍♀️

**Comment by kkostov** [comment]
Yes, that's not an easy one. A second type sounds like a good idea, but then we specialise "time based" collections even though we still don't have guarantees about the sort order (previous/next can still be reversed).

How about this - what if we try to keep one type (`PagedResult`) but assume no knowledge about the ordering of items e.g. we focus on the directionality relative to the current position in the collection (and highlight that we don't know what the ordering system is, because it's server-driven). e.g. 

* `primaryCursor` (`canMovePrimary`): Moves in the direction provided by the server, which could mean fetching newer items, earlier items in a list, etc.
* `secondaryCursor` (`canMoveSecondary`): Moves in the opposite direction, which could mean fetching older items, later items in a list, etc.

```swift
if posts.canMovePrimary, let morePosts = posts.primaryCursor() {
    // Load more posts in the server-provided direction (e.g., newer posts)
}

if posts.canMoveSecondary, let morePosts = posts.secondaryCursor() {
    // Load posts in the opposite direction (e.g., older posts)
}
```

```swift
if bookmarks.canMovePrimary, let moreBookmarks = bookmarks.primaryCursor() {
    // Load more bookmarks from the list (historically older perhaps)
}

if bookmarks.canMoveSecondary, let moreBookmarks = bookmarks.secondaryCursor() {
    // Load more bookmarks in the opposite direction
}
```

Perhaps @davidgarywood can also jump in, I know he has good naming suggestions and a way of keeping us from making it even more confusing 😅.

**Comment by davidgarywood** [comment]
A few things here that spring to my mind. Just to collect context so far:

* @helje5 has run head first into this, and @technicat has a wrapper but is used to the behaviour, so we should do something.

* We cannot guarantee the ordering, even if _generally_ timelines behave one way, and bookmarks etc another.

* The API and server don’t tell us which way round things are operating.

So then I wonder, what does it look like for people at point of use?

Could we take an approach like this:
* Document the issue to help avoid confusion.
* Hoist the complexity up a level into something new that sits on top? (This would have the added bonus of not breaking existing compatibility)
* If so, what is the commonality in point of use, or a typical use case? Do people care about cursors or position, or are folks typically starting from 0, and then advancing until they have everything, with some then caching what they’ve had already?

That last bullet point probably being key to threading the needle 😅

**Comment by helje5** [comment]
> The API and server don’t tell us which way round things are operating.

I don't have a grasp of the full API exposed by this project, but the HTTP `Link` header values certainly have a fixed meaning of direction and are unrelated to endpoints (or should be, maybe Masto got this wrong?). Which is getting the next batch and unrelated to any timing concepts in the payloads.
That's why I thought that maybe two types for two different things may be proper.

**Comment by daprice** [comment]
I think the current TootSDK abstraction is plenty, the main problem is the “previous” and “next” names are (a) the opposite of what Mastodon calls them, and (b) difficult to reason about regardless of which way around they are, because we aren’t necessarily dealing with time-based ordering.

But what we *are* always dealing with is ID-based ordering, with what masto calls “prev” always being **higher IDs** and what it calls “next” being **lower IDs** (at least I think that’s the right way around 😰). So I’d suggest simply renaming TootSDK’s previous and next page properties to `ascendingPage` (higher IDs) and `descendingPage` (lower IDs).

This is semantically accurate, doesn’t directly conflict with masto’s terminology (I know I keep having to remind myself which is which any time I jump between masto’s docs and my code), and has the nice side effect of spatially mapping to the way posts are typically presented in UI – scroll up for the ascending page and down for the descending page.

<details><summary>re: use cases</summary>
<p>

> If so, what is the commonality in point of use, or a typical use case? Do people care about cursors or position, or are folks typically starting from 0, and then advancing until they have everything, with some then caching what they’ve had already?

I think there are some cases where I do just want to start at the “first” page and advance until I have everything, like when fetching the primary account’s following, blocked users, or domain blocks for example. But especially for timelines, I definitely need to care about paging in both directions:

- when loading a new timeline from scratch, start at the “first” page and load more as the user scrolls
- when refreshing a timeline, load newer pages relative to the previous “first” page until we arrive at the most recent posts
- when restoring reading position (e.g. from the mastodon marker API), if the last read post ID isn’t already loaded, fetch the pages immediately before and after the last read post ID if they haven’t dropped off the home timeline, and fill in any gaps between them and whatever else has already been loaded
- if the app is using websocket streaming, anytime streaming is ended I have to synthesize a new “newest page” based on what was received from streaming, so page-based fetching can “catch up” from where the stream left off
- …etc.


</p>
</details>

**Comment by kkostov** [comment]
I suggest we stick to updating the docs for `PagedResult` to better explain what "previous" and "next" means. I also really like the `getPages` wrapper of @technicat which sounds like a good addition. It's easy to add and works in parallel with the existing approach. We can update the docs to focus on using `getPages` so folks will only get confronted with previous/next when they manually need to navigate the collection.

**Comment by technicat** [comment]
re `getPages`, the `Pageable` protocol can just be replaced with `Decodable` to satisfy `PagedResult`. `Pageable` also includes `Identifiable` and `Equatable` because I use it in the UI for scrolling pages, but I just realized I don't have to use the same query signature for both. The UI code looks like this:

```swift
LazyVStack(alignment: .leading) {
      ForEach(list) { item in
        row(item)
          .onAppear {
            if item == list.last, prev != nil {
              status.run { try await getPrev() }
            }
          }
      }
    }
```
